packages/amplify-cli-core/src/types.ts

```diff
@@ -20,19 +20,209 @@ export type $TSContext = {
+  executable: string;
+  category: string;
+  isHeadless: boolean;
+  cmd: string;
+  subCmd: string | undefined;
+  optionFlowData: Array<TypeOptionFlowData>; //IOptionFlowHeadlessData | IOptionFlowCLIData
+  input: $CommandLineInput;
+  timestamp: string;
+  projectEnvIdentifier?: string; // hash(ProjectName + Amplify AppId + EnvName)
+  projectIdentifier?: string; // hash( ProjectName + Amplify App Id)
+}
+export type TimedCodePath = ManuallyTimedCodePath | UntilExitTimedCodePath | FromStartupTimedCodePaths;
+export interface IUsageMetricsData {
+  emitAbort: () => Promise<void>;
+  emitError: (error: Error | null) => Promise<void>;
+  emitSuccess: () => Promise<void>;
+  init: (
+    installationUuid: string,
+    version: string,
+    input: $CommandLineInput,
+    accountId: string,
+    projectSettings: ProjectSettings,
+    processStartTimeStamp: number,
+  ) => void;
+  getUsageDataPayload: (error: Error | null, state: string) => IUsageDataPayload;
+  startCodePathTimer: (codePath: StartableTimedCodePath) => void;
+  stopCodePathTimer: (codePath: StoppableTimedCodePath) => void;
+  calculatePushNormalizationFactor: (events: { StackId: string; PhysicalResourceId: string }[], StackId: string) => void;
+  getSessionUuid: () => string;
+}
+
+export type StartableTimedCodePath = ManuallyTimedCodePath | UntilExitTimedCodePath;
+export type StoppableTimedCodePath = ManuallyTimedCodePath | FromStartupTimedCodePaths;
+
+export enum FromStartupTimedCodePaths {
+  PLATFORM_STARTUP = 'platformStartup', // time from CLI process start to plugin invoke. This timer is auto started on process startup
+  TOTAL_DURATION = 'totalDuration', // time from CLI process start to exit
+}
+export enum UntilExitTimedCodePath {
+  POST_PROCESS = 'postProcess', // time from plugin exit to process exit. This timer is automatically stopped at the end of the process
+}
+export enum ManuallyTimedCodePath {
+  // general paths (applies to all commands)
+  PLUGIN_TIME = 'pluginTime', // time spent in command handler plugin
+
+  // push-specific paths
+  PUSH_TRANSFORM = 'pushTransform', // total time spent transforming resources and uploading assets to prepare for a push
+  PUSH_DEPLOYMENT = 'pushDeployment', // time spent deploying CFN
+
+  // init-specific paths (also called during env checkout and pull)
+  INIT_ENV_PLATFORM = 'initEnvPlatform', // time to call awscloudformation provider initEnv. This includes downloading deployment bucket and updating local files
+  INIT_ENV_CATEGORIES = 'initEnvCategories', // time to call all of the category's initEnv methods
+
+  PROMPT_TIME = 'promptTime', // total time to takes to answer a prompt
+}
+
+export interface ContextParameters extends Pick<$CommandLineInput, 'argv' | 'plugin' | 'command' | 'options'> {
+  raw: $CommandLineInput['argv'];
+  array: $CommandLineInput['subCommands'];
+  first?: string;
+  second?: string;
+  third?: string;
+}
+
+export type $CommandLineInput = {
+  argv: Array<string>;
+  plugin?: string;
```

Comment:
```suggestion
  pluginName?: string;
```

Reply:
I think it's best to not change this right now. This is one of those words that is used everywhere in the project. This refactor should be done in its own PR.